server/services/src/main/java/org/zanata/service/impl/GlossaryFileServiceImpl.java

```diff
@@ -116,8 +119,40 @@ public GlossaryProcessed saveOrUpdateGlossary(
                 }
                 continue;
             }
-            message = checkForDuplicateEntry(entry);
+
             boolean onlyTransferTransTerm = false;
+
+            if (StringUtils.isBlank(entry.getExternalId())) {
+                boolean success = false;
+                Optional<GlossaryTerm> srcTerm = getSourceTerm(entry);
+                if (!srcTerm.isPresent()) {
+                    continue;
+                }
+                // Assume first import, give it a new ID based on term and pos
+                String newId = srcTerm.get().getContent().concat("-").concat(entry.getPos());
+
+                entry.setExternalId(newId);
+                if (entryExists(entry)) {
```

Comment:
@aeng @seanf I'm sure I'm doing something evil here. Opinions?

Reply:
What is that id for? would be better to use hash on those fields?

Reply:
It's for external management, so terms can be uploaded with an id for tracking and updating. e.g.
```
{ terms:[{
  id: "thing-noun",
  term: "thing",
  description: "a whatsit"
}] }
```
This allows for instances where you have multiple definitions, like "bat" having multiple nouns and verbs, and so that if you fix a typo in a description you don't end up with duplicate entries.
(This was a side project of mine, so I should make a jira that properly explains it.)

Reply:
In texflow, we use hash as the external resid. I think we should do the same here. Hashing of the content and pos

Reply:
The problem is you can have e.g. multiple "bat" terms that are nouns, which will incorrectly resolve to the same hash. Further, you can change the term's POS or Description in the glossary editor and reimport the same glossary file, resulting in duplication - or do this externally with the same result.

Reply:
Would it be more sensible to get a list of every id that matches "term-pos*" and match + increment locally?

Reply:
Yes and that has always been the problem with our glossary. That is why we use the db:id instead. The hash column is used for checking when new glossary entry is being created (on server side).